sqlglot/optimizer/qualify_columns.py

```diff
@@ -0,0 +1,321 @@
+    selectables = _merge(
+        selectables,
+        _qualify_derived_tables(derived_tables, schema, sequence, parent_selectables),
+    )
+    expansions = _expand_stars(select_stars, selectables)
+    columns.extend(expansions)
+    selections.extend(expansions)
+    selectables = _merge(parent_selectables, selectables)
+    _qualify_columns(columns, selectables)
+    _qualify_subqueries(subqueries, schema, sequence, selectables)
+    return _qualify_outputs(selections)
+
+
+def _qualify_derived_tables(
+    derived_tables, schema, sequence, parent_selectables, chain=False
+):
+    selectables = {}
+    for subquery in derived_tables:
+        subquery_outputs = _qualify_union(
+            expression=subquery.this,
+            schema=schema,
+            sequence=sequence,
+            parent_selectables=_merge(parent_selectables, selectables)
+            if chain
+            else parent_selectables,
+        )
+
+        table_alias = subquery.args.get("alias")
+        if not table_alias:
+            table_alias = exp.TableAlias()
+            subquery.set("alias", table_alias)
+
+        alias = table_alias.args.get("this")
+        if not alias:
+            alias = exp.to_identifier(f"_alias{sequence.next()}")
+            table_alias.set("this", alias)
+
+        alias_columns = table_alias.args.get("columns")
+        if not alias_columns:
+            alias_columns = [exp.to_identifier(o) for o in subquery_outputs]
+            table_alias.set("columns", alias_columns)
+
+        selectable_table = alias.text("this")
+
+        selectable_columns = [c.text("this") for c in alias_columns]
+        selectables = _merge(selectables, {selectable_table: selectable_columns})
+
+    return selectables
+
+
+def _qualify_tables(tables, schema, parent_selectables):
+    selectables = {}
+    for table in tables:
+        table_name = table.text("this")
+
+        if isinstance(table.parent, exp.Alias):
+            selectable_table = table.parent.alias
+        else:
+            selectable_table = table_name
+
+        if table_name in parent_selectables:
+            selectable_columns = parent_selectables[table_name]
+        else:
+            selectable_columns = list(schema[table_name])
+
+        selectables = _merge(selectables, {selectable_table: selectable_columns})
+    return selectables
+
+
+def _expand_stars(select_stars, selectables):
+    result = []
+    for star in select_stars:
+        new_columns = []
+
+        if isinstance(star, exp.Column):
+            tables = [star.text("table")]
+        else:
+            tables = list(selectables)
+
+        for table in tables:
+            columns = selectables.get(table, [])
+            for column in sorted(columns):
+                new_columns.append(
+                    exp.Column(
+                        this=exp.to_identifier(column), table=exp.to_identifier(table)
+                    )
+                )
+        star.replace(*new_columns)
```

Comment:
i've been looking for something like this, you're able to replace a single child node with multiple?

sort of unrelated to this PR, but can you add this functionality to def transform? so if i return an array it works? 

if i'm understanding correctly, then this is great!

Reply:
Yeah, it does exactly that. I think this already works for `transform` in the way I wrote this. I'll add a test to prove it.